### PR TITLE
[POC RFC] header only API for singletons

### PR DESCRIPTION
**Pull request by marcalff**
Fix header only API for singletons (#1520)

## Changes

DRAFT -- implement header only API singletons.

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by lalitb** [comment]
Did a quick test with this PR on Linux, by creating an instrumented library with `-fvisible=hidden` flag and the library is able to use the singleton tracer provider from the application. So changes seem to work on Linux.

Need to check on Windows, but based on what @owent has already tested [here](https://github.com/open-telemetry/opentelemetry-cpp/issues/1520#issuecomment-1198104171), it may not work with PE ABI as `__declspec(dllexport)` needs to be held inside the compile unit.

**Comment by owent** [comment]
I have add more examples [here](https://github.com/open-telemetry/opentelemetry-cpp/issues/1520#issuecomment-1198104171), and `__declspec (selectany)` of MSVC works well, but `__attribute__ ((selectany))` of GCC on Windows do not work as expected,do I miss anything?

**Comment by marcalff** [comment]
Paper from Ulrich Drepper, see section 2.2 export control, page 17:

https://akkadia.org/drepper/dsohowto.pdf

**Comment by owent** [comment]
> Resources:
> 
> 
> Paper from Ulrich Drepper, see section 2.2 export control, page 17:
> 
> https://akkadia.org/drepper/dsohowto.pdf
> 
> 
> Gnu visibility doc
> 
> https://gcc.gnu.org/wiki/Visibility

I have a quick look at https://akkadia.org/drepper/dsohowto.pdf, I think it's for ELF ABI of Linux, not PE ABI of Windows.
And the `visibility` attribute of gcc document(https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#Common-Function-Attributes) also says ***The possible values of visibility_type correspond to the visibility settings in the ELF gABI.*** .

My question is: is there any way to make the singleton unique with GCC on Windows?

**Comment by lalitb** [comment]
@owent - We agreed this PR to be scoped for Linux and Mac. Let us know if you have any comment on that.

**Comment by owent** [comment]
> @owent - We agreed this PR to be scoped for Linux and Mac. Let us know if you have any comment on that.

Fine by me. I think we can document this if we do not support build shared library with gcc and clang on Windows, can we also force set `BUILD_SHARED_LIB` to `OFF` when we use gcc and clang on Windows to reduce misuse?

@lalitb @marcalff

**Review by owent — APPROVED** [review]
LGTM, Do we need set `BUILD_SHARED_LIB` to `OFF` with gcc on Windows to avoid misuse?

**Comment by marcalff** [comment]
@lalitb , @esigo , @owent 

I added a singleton_test unit test, and makefiles for both CMake and Bazel.

This test does not build properly on windows platforms, and I can not figure out why.

"component_c" is compiled as a shared library / DDL (expected),
but then when linking the final singleton_test binary, the makefile wants a component_c.lib instead of component_c.ddl,
which is incorrect.

On linux, the same makefile with cmake properly builds static libraries (component_a, component_b) and shared libraries (component_c, d, e and f) as expected.

For bazel, I could not find a proper way to express in the BUILD file how to link a test binary singleton_test with a mix of static (component_a and b) and shared (component_c, d, e and f) libraries.

Any help appreciated, the goal is to have first a unit test that builds everywhere.

Regards,
-- Marc

File: api/include/opentelemetry/metrics/provider.h [hunk 1]
```diff
@@ -29,8 +44,8 @@ class Provider
    */
   static nostd::shared_ptr<MeterProvider> GetMeterProvider() noexcept
   {
-    std::lock_guard<common::SpinLockMutex> guard(lock);
-    nostd::shared_ptr<MeterProvider> result(provider);
+    std::lock_guard<common::SpinLockMutex> guard(s.lock);
```
Comment (owent) [inline api/include/opentelemetry/metrics/provider.h:47]: I think we should not use a short name here

**Comment by owent** [comment]
> @lalitb , @esigo , @owent
> 
> I added a singleton_test unit test, and makefiles for both CMake and Bazel.
> 
> This test does not build properly on windows platforms, and I can not figure out why.
> 
> "component_c" is compiled as a shared library / DDL (expected), but then when linking the final singleton_test binary, the makefile wants a component_c.lib instead of component_c.ddl, which is incorrect.
> 
> On linux, the same makefile with cmake properly builds static libraries (component_a, component_b) and shared libraries (component_c, d, e and f) as expected.
> 
> For bazel, I could not find a proper way to express in the BUILD file how to link a test binary singleton_test with a mix of static (component_a and b) and shared (component_c, d, e and f) libraries.
> 
> Any help appreciated, the goal is to have first a unit test that builds everywhere.
> 
> Regards, -- Marc

Do we need declare `do_something_in_c` as default visibility or `__declspec(dllexport)/__declspec(dllimport)` ? I find `component_c` is built as a shared library but export nothing.

To my knowledge, MSVC will create both `.lib` and `.dll` for dynamic libraries. `.lib` is for linking and `.dll` is for runtime.

**Comment by marcalff** [comment]
> Do we need declare `do_something_in_c` as default visibility or `__declspec(dllexport)/__declspec(dllimport)` ? I find `component_c` is built as a shared library but export nothing.

Good point, fixing this.

Thanks

**Comment by marcalff** [comment]
This PR contains proof of concept code, for a general fix.

It is now closed, for the benefit of:
- PR #1604, a fix for gcc and clang, ready for merge
- PR #1595, a tentative fix for windows, still a proof of concept.

Work for windows should continue in PR #1595.

**Pull request closed without merging**